File: solutions/0301-0400/0350-intersection-of-two-arrays-ii/solution.cpp
#include "solution.hpp"

#include <unordered_map>

namespace leetcode::p0350 {

std::vector<int> solve(const std::vector<int>& nums1, const std::vector<int>& nums2) {
    std::unordered_map<int, int> counts;
    for (int value : nums1) {
        ++counts[value];
    }

    std::vector<int> result;
    for (int value : nums2) {
        auto it = counts.find(value);
        if (it != counts.end() && it->second > 0) {
            result.push_back(value);
            --it->second;
        }
    }

    return result;
}

} // namespace leetcode::p0350

File: solutions/0301-0400/0350-intersection-of-two-arrays-ii/solution.ml
let solve nums1 nums2 =
  let counts = Hashtbl.create 16 in
  List.iter
    (fun n ->
      let current = Option.value (Hashtbl.find_opt counts n) ~default:0 in
      Hashtbl.replace counts n (current + 1))
    nums1;
  let step (acc, tbl) n =
    match Hashtbl.find_opt tbl n with
    | Some count when count > 0 ->
        Hashtbl.replace tbl n (count - 1);
        (n :: acc, tbl)
    | _ -> (acc, tbl)
  in
  fst (List.fold_left step ([], counts) nums2)
